test/Microsoft.AspNetCore.Authentication.Test/CollectingEventListener.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Authentication.Test
+{
+    // REVIEW: Also to go to some testing or shared-source assembly
+    // This can only collect from EventSources created AFTER this listener is created
```

Comment:
Is this a **lie** then?

>Called for all existing event sources when the event listener is created and when a new event source is attached to the listener.

https://msdn.microsoft.com/en-us/library/system.diagnostics.tracing.eventlistener.oneventsourcecreated(v=vs.110).aspx

Reply:
The comment is out of date. The problem is that the event is called during the base class constructor which means that you can't access any constructor arguments (making it non-trivial to dynamically attach to event sources created before the listener was created by name). We were able to refactor around some locks to allow it to collect events from sources created before the listener was created.

Reply:
ah, the good old fragile base class problem